python/python/bystro/tada_mixture/mixture_tada_marginal.py

```diff
@@ -0,0 +1,214 @@
+"""
+This maximizes the marginal likelihood for the Poisson model, for now only 
+a Possion is implemented. The generative model is 
+
+p(z) = Dir(alpha_k)
+p(x_ij|z=k) = Poisson(Lambda_{jk})
+
+The log likelihood is 
+
+log p(x) = log(\sum p(z=k)p(x_ij|z=k)). 
+
+There's not an analytic formula that makes this easy to solve. But as long
+as you use the logsumexp trick you can compute this easily numerically. I 
+honestly don't know why that isn't taught more, I only learned it as a 
+postdoc even though I got a PhD in statistics. Meh
+
+
+Objects
+-------
+MVTadaPoissonML(K=4,training_options={})
+    This fits a standard Poisson latent variable model.
+
+Methods
+-------
+None
+
+"""
+import numpy as np
+import torch
+from tqdm import trange  # type: ignore
+from torch import nn
+from torch.distributions import Dirichlet
+import scipy.stats as st  # type: ignore
+from sklearn.mixture import GaussianMixture  # type: ignore
+from torch.nn import PoissonNLLLoss
+
+
+class MVTadaPoissonML(object):
+    def __init__(self, K=4, training_options=None):
+        """
+        This is a product Poisson latent variable model.
+
+        Parameters
+        ----------
+        K : int,default=4
+            The number of clusters, i.e. number of different types of 
+            genes
+
+        training_options : dict,default={}
+            The parameters for the inference scheme
```

Comment:
How serious of a problem is this? Is there a standardized way we should address this?

Reply:
In my experience with large python codebases, the use of dictionaries to represent records (especially config objects) is one of the largest single sources of programming errors, because the keys can't be statically checked.  The resulting bug doesn't cause a runtime error, it just silently ignores the caller's intent.  Depending on how many kwargs a class needs to take upon initialization, they can either be passed directly (ideally in kwarg-only position), with a TypedDict (https://docs.python.org/3.8/library/typing.html#typing.TypedDict), or in a separate config dataclass.